lib/liquid_voting_web/resolvers/delegations.ex

```diff
@@ -8,27 +8,109 @@ defmodule LiquidVotingWeb.Resolvers.Delegations do
   def delegation(_, %{id: id}, %{context: %{organization_id: organization_id}}),
     do: {:ok, Delegations.get_delegation!(id, organization_id)}
 
-  # Will add participants to the db if they don't exist yet, or fetch them if they do. 
-  # Their ids are used for delegator_id and delegate_id when inserting the delegation
-  # with create_delegation_with_valid_arguments/1
+  @doc """
+  Creates a delegation.
+
+  Adds participants to the db if they don't exist, or fetches them if they do.
+
+  Valid arguments (args) must include 2 ids (delegator_id and delegate_id)
+  OR 2 emails (delegator_email and delegate_email).
+
+  Valid arguments (args) may include a proposal_url.
+  Without a proposal_url, an attempt to create a global delegation will occur.
+
+  The participant ids, either directly taken from delegator_id and
+  delegate_id, or via searching the db for participants with the emails
+  provided, are used when inserting the delegation with
+  LiquidVoting.Delegations.create_delegation/1.
+
+  ## Examples
+
+  iex> create_delegation(
+    %{},
+    %{delegator_email: "alice@somemail.com",
+    delegate_email: "bob@somemail.com",
+    proposal_url: "https://proposalplace/proposal63"},
+    %{context: %{organization_id: "b212ef83-d3df-4a7a-8875-36cca613e8d6"}})
+  %Delegation{}
+
+  iex> create_delegation(
+    %{},
+    %{delegator_email: "alice@somemail.com"},
+    %{context: %{organization_id: "b212ef83-d3df-4a7a-8875-36cca613e8d6"}})
+  {:error,                                                   
+    %{
+      details: %{delegate_email: ["can't be blank"]},
+      message: "Could not create delegation"
+    }}
+  """
   def create_delegation(_, args, %{context: %{organization_id: organization_id}}) do
     args
-    |> Map.put(:organization_id, organization_id)
-    |> Delegations.create_delegation()
+    |> validate_participant_args
     |> case do
```

Comment:
Suggestion: refactor this to use with, something like this: 

```elixir
  with {:ok, args} <- validate_participant_args(args),
     {:ok, delegation} <- Delegations.create_delegation(args) do
  {:ok, delegation}
else
  error -> error
end
```

(`Map.put(args, :organisation_id)` could be done beforehand)

Reply:
I like this idea, but how do I include the 3 different types of error tuples that could be returned (1 type from the call to `validate_participant()` and 2 different types possible from call to `create_delegation()`) in the `with` structure you suggest?  
There seems to be only 1 error case, in the `else` block, whilst we could errors for either call.

Reply:
You could move the three errors associated with `create_delegation` to the else block and bubble up the ones from `validate_participant` with the catch-all.

Reply:
Do you mean this kind of approach?  
```
def convert_to_csv(source_filename) do
    with {:l1, {:ok, :xlsx}} <- {:l1, detect_file_type(source_filename)},
         {:l2, {:ok, csv_filename}} <- {:l2, convert_excel_to_csv(source_filename)} do
  {:ok, csv_filename}
else
  {:l1, {:error, error}} -> {:error, error}  # (such as file not found)
  {:l1, {:ok, _other}} -> {:error, :not_an_excel_file}
  {:l2, error} -> error
end
```

Reply:
Something else I've noticed: we have two different forms for the error messages. Might be nice to homogenise this (later PR of course, but worth discussing, I think)?

Reply:
> Do you mean this kind of approach?

Essentially you just need to move those errors for `create_delegation` somewhere, and I think the else block is fine as we don't want to introduce another wrapper around that function.

Reply:
Agree. Errors need some love. Maybe create an issue?

Reply:
Ah my bad, you don't need to wrap the structs in steps. I was thinking: 

```elixir
 with {:ok, args} <- validate_participant_args(args),
     {:ok, delegation} <- Delegations.create_delegation(args) do
  {:ok, delegation}
else
  {:error, changeset} -> handle()
  {:error, name, changeset, _} -> handle()
  error -> error
end
```

Though I now realise that for this to work, you'd need to change the _shape_  of the errors in the validation function to `{:error, message, detail}`
```

Reply:
OK, I can try that.  

BTW, I based the error structure on `assert details == %{email: ["is invalid"]}` from `test/liquid_voting_web/absinthe/mutations/create_participant_test.exs`

Reply:
I see, could you add some detail to the error situation in #123?

Reply:
To be clear, what are you seeing as the default / desired error structure?  Is it this:  
`{:ok, %{errors: [%{message: message, details: details}]}}`

Reply:
I don't know what the default or desired structure is, that's the reason for opening #123. However, the structure we've had in the resolvers up until now is `{:error, message, detail}`.

Reply:
I'm disappearing down a bit of a rabbit hole here, trying to get the errors to play well with the error -> error part, since create_delegation() can also return errors from upsert_participant, and these errors are not all formatted the way you are envisioning. Thus, I would need to change the error shapes elsewhere (as in the example above, which seems to be changeset error), which seems like scope creep.   
How about we go with the steps wrappers for now, and simplify after the errors issue is resolved?

Reply:
In that case I'd revert to the nested case statements for now, we don't want the added complexity of steps in our return values. Let's save the rabbit-hole for another day once #123 is discussed.

Reply:
Actually, this almost works...
```
with {:ok, args} <- validate_participant_args(args),
         {:ok, delegation} <- Delegations.create_delegation(args) do
      {:ok, delegation}
    else
      {:error, message} ->
        {:error, message}

      {:error, changeset} ->
        {:error,
         message: "Could not create delegation", details: ChangesetErrors.error_details(changeset)}

      {:error, name, changeset, _} ->
        {:error,
         message: "Could not create #{name}", details: ChangesetErrors.error_details(changeset)}
    end
```
Just getting one (other) test throwing an error now.

Reply:
@niklaslong I think I finally have a simpler version, in line with your ideas, but still deferring reworking the error formats until another issue. Thanks for helping me with this, and especially with understand the matching issue!